From f805944eec10cd2f8c37c98d1232601b50ebaee9 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Wed, 26 Nov 2014 00:53:39 -0500
Subject: [PATCH] Fix EAR AddToChunk and fireworks bug

We should never skip a tick if the entity is flagged as "not added to chunk" so it can be added correctly.
Also never skip fireworks or projectiles
---
 src/main/java/net/minecraft/server/Entity.java  | 2 +-
 src/main/java/org/spigotmc/ActivationRange.java | 5 +++++
 2 files changed, 6 insertions(+), 1 deletion(-)

diff --git a/src/main/java/net/minecraft/server/Entity.java b/src/main/java/net/minecraft/server/Entity.java
index 153337b..fd07ca2 100644
--- a/src/main/java/net/minecraft/server/Entity.java
+++ b/src/main/java/net/minecraft/server/Entity.java
@@ -111,7 +111,7 @@ public abstract class Entity {
     protected DataWatcher datawatcher;
     private double g;
     private double h;
-    public boolean ag;
+    public boolean ag;public boolean isAddedToChunk() { return ag; } // Spigot
     public int ah;
     public int ai;
     public int aj;
diff --git a/src/main/java/org/spigotmc/ActivationRange.java b/src/main/java/org/spigotmc/ActivationRange.java
index f69cc65..303f71c 100644
--- a/src/main/java/org/spigotmc/ActivationRange.java
+++ b/src/main/java/org/spigotmc/ActivationRange.java
@@ -263,6 +263,11 @@ public class ActivationRange
     {
         if (entity.isDisabled) return true; // EMC
         SpigotTimings.checkIfActiveTimer.startTiming();
+        // Never safe to skip fireworks or entities not yet added to chunk
+        if (!entity.isAddedToChunk() || entity instanceof EntityFireworks || entity.getBukkitEntity() instanceof org.bukkit.entity.Projectile) {
+            SpigotTimings.checkIfActiveTimer.stopTiming();
+            return true;
+        }
         boolean isActive = entity.activatedTick >= MinecraftServer.currentTick || entity.defaultActivationState;
 
         // Should this entity tick?
-- 
1.9.1

